### PR TITLE
Add FIM tests for recursion_level

### LINKED ISSUE #146: Syscheck automated tests: 'recursion_level' option
| Working branch |
| --- |
|[fim-integration-tests-recursion](https://github.com/wazuh/wazuh-qa/tree/fim-integration-tests-recursion)|

## Description

This issue is part of the roadmap in #7.

The `maximum_recursion_level` tag is used to configure the maximum recursion level allowed for a specific directory so `syscheck` will generate alerts for all files up to the indicated directory level but will ignore the rest.

According to [documentation](https://documentation.wazuh.com/3.10/user-manual/capabilities/file-integrity/fim-configuration.html#configure-maximum-recursion-level-allowed), using the following directory structure and `recursion_level="3"`:

```
folder_test
├── file_0.txt
└── level_1
    ├── file_1.txt
    └── level_2
        ├── file_2.txt
        └── level_3
            ├── file_3.txt
            └── level_4
                ├── file_4.txt
                └── level_5
                    └── file_5.txt
```
We will receive alerts for all files up to `folder_test/level_1/level_2/level_3/` but we won’t receive alerts from any directory deeper than level_3.

The `maximum_recursion_level` must be an integer *between 0 and 320*.

## Subtasks
- Realtime
  - [x] Monitor and ignore files with Default Recursion Level
  - [x] Monitor one folder and ignore all subdirectories with Recursion Level 0
  - [x] Monitor one folder and its level 1 subdirectory but ignore the rest with Recursion Level 1
  - [x] Monitor one folder and its subdirectories until level 5 with Recursion Level 5
  - [x] Monitor files on the deepest subdirectory level with recursion level 320
- Whodata
  - [x] Monitor and ignore files with Default Recursion Level
  - [x] Monitor one folder and ignore all subdirectories with Recursion Level 0
  - [x] Monitor one folder and its level 1 subdirectory but ignore the rest with Recursion Level 1
  - [x] Monitor one folder and its subdirectories until level 5 with Recursion Level 5
  - [x] Monitor files on the deepest subdirectory level with recursion level 320
- Scheduled
  - [x] Monitor and ignore files with Default Recursion Level
  - [x] Monitor one folder and ignore all subdirectories with Recursion Level 0
  - [x] Monitor one folder and its level 1 subdirectory but ignore the rest with Recursion Level 1
  - [x] Monitor one folder and its subdirectories until level 5 with Recursion Level 5
  - [x] Monitor files on the deepest subdirectory level with recursion level 320

### PR DESCRIPTION
**Pull request by CarlosRS9**
Hi Team,

This closes #146. 

I created tests for the `recursion_level` option for different wazuh configurations. To achieve this, among with the creation of the complete set of tests, I had to made some changes over `regular_file_cud` in `fim.py`:

- The function now accepts one extra parameter (optional) to be able to use it in tests that are meant to fail as this will indicate if the event should be triggered or not (for example: check if creating a file outside the recursion_level range triggers an event).

- The logic corresponding to this new functionality has been implemented inside the function.

- Further improvements are made on the body of this as several parts were used more than once.

As this change may have impact over past functionality all previous tests have been executed and completed as expected.

### Tests example

```
================================================= test session starts =================================================
platform linux -- Python 3.6.8, pytest-5.2.0, py-1.8.0, pluggy-0.13.0 -- /bin/python3
cachedir: .pytest_cache
rootdir: /vagrant/wazuh-qa/test_wazuh, inifile: pytest.ini
collected 72 items

test_recursion_level.py::test_recursion_realtime[get_configuration0-/test_no_recursion-subdir-0] SKIPPED         [  1%]
test_recursion_level.py::test_recursion_realtime[get_configuration0-/test no recursion-sub dir -0] SKIPPED       [  2%]
test_recursion_level.py::test_recursion_realtime[get_configuration0-/test_recursion_1-subdir-1] SKIPPED          [  4%]
test_recursion_level.py::test_recursion_realtime[get_configuration0-/test recursion 1-sub dir -1] SKIPPED        [  5%]
test_recursion_level.py::test_recursion_realtime[get_configuration0-/test_recursion_5-subdir-5] SKIPPED          [  6%]
test_recursion_level.py::test_recursion_realtime[get_configuration0-/test recursion 5-sub dir -5] SKIPPED        [  8%]
test_recursion_level.py::test_recursion_realtime[get_configuration0-/test_recursion_320-subdir-318] SKIPPED      [  9%]
test_recursion_level.py::test_recursion_realtime[get_configuration0-/test recursion 320-sub dir -318] SKIPPED    [ 11%]
test_recursion_level.py::test_recursion_scheduled[get_configuration0-/test_no_recursion-subdir-0] SKIPPED        [ 12%]
test_recursion_level.py::test_recursion_scheduled[get_configuration0-/test no recursion-sub dir -0] SKIPPED      [ 13%]
test_recursion_level.py::test_recursion_scheduled[get_configuration0-/test_recursion_1-subdir-1] SKIPPED         [ 15%]
test_recursion_level.py::test_recursion_scheduled[get_configuration0-/test recursion 1-sub dir -1] SKIPPED       [ 16%]
test_recursion_level.py::test_recursion_scheduled[get_configuration0-/test_recursion_5-subdir-5] SKIPPED         [ 18%]
test_recursion_level.py::test_recursion_scheduled[get_configuration0-/test recursion 5-sub dir -5] SKIPPED       [ 19%]
test_recursion_level.py::test_recursion_scheduled[get_configuration0-/test_recursion_320-subdir-318] SKIPPED     [ 20%]
test_recursion_level.py::test_recursion_scheduled[get_configuration0-/test recursion 320-sub dir -318] SKIPPED   [ 22%]
test_recursion_level.py::test_recursion_whodata[get_configuration0-/test_no_recursion-subdir-0] SKIPPED          [ 23%]
test_recursion_level.py::test_recursion_whodata[get_configuration0-/test no recursion-sub dir -0] SKIPPED        [ 25%]
test_recursion_level.py::test_recursion_whodata[get_configuration0-/test_recursion_1-subdir-1] SKIPPED           [ 26%]
test_recursion_level.py::test_recursion_whodata[get_configuration0-/test recursion 1-sub dir -1] SKIPPED         [ 27%]
test_recursion_level.py::test_recursion_whodata[get_configuration0-/test_recursion_5-subdir-5] SKIPPED           [ 29%]
test_recursion_level.py::test_recursion_whodata[get_configuration0-/test recursion 5-sub dir -5] SKIPPED         [ 30%]
test_recursion_level.py::test_recursion_whodata[get_configuration0-/test_recursion_320-subdir-318] SKIPPED       [ 31%]
test_recursion_level.py::test_recursion_whodata[get_configuration0-/test recursion 320-sub dir -318] SKIPPED     [ 33%]
test_recursion_level.py::test_recursion_realtime[get_configuration1-/test_no_recursion-subdir-0] PASSED          [ 34%]
test_recursion_level.py::test_recursion_realtime[get_configuration1-/test no recursion-sub dir -0] PASSED        [ 36%]
test_recursion_level.py::test_recursion_realtime[get_configuration1-/test_recursion_1-subdir-1] PASSED           [ 37%]
test_recursion_level.py::test_recursion_realtime[get_configuration1-/test recursion 1-sub dir -1] PASSED         [ 38%]
test_recursion_level.py::test_recursion_realtime[get_configuration1-/test_recursion_5-subdir-5] PASSED           [ 40%]
test_recursion_level.py::test_recursion_realtime[get_configuration1-/test recursion 5-sub dir -5] PASSED         [ 41%]
test_recursion_level.py::test_recursion_realtime[get_configuration1-/test_recursion_320-subdir-318] PASSED       [ 43%]
test_recursion_level.py::test_recursion_realtime[get_configuration1-/test recursion 320-sub dir -318] PASSED     [ 44%]
test_recursion_level.py::test_recursion_scheduled[get_configuration1-/test_no_recursion-subdir-0] SKIPPED        [ 45%]
test_recursion_level.py::test_recursion_scheduled[get_configuration1-/test no recursion-sub dir -0] SKIPPED      [ 47%]
test_recursion_level.py::test_recursion_scheduled[get_configuration1-/test_recursion_1-subdir-1] SKIPPED         [ 48%]
test_recursion_level.py::test_recursion_scheduled[get_configuration1-/test recursion 1-sub dir -1] SKIPPED       [ 50%]
test_recursion_level.py::test_recursion_scheduled[get_configuration1-/test_recursion_5-subdir-5] SKIPPED         [ 51%]
test_recursion_level.py::test_recursion_scheduled[get_configuration1-/test recursion 5-sub dir -5] SKIPPED       [ 52%]
test_recursion_level.py::test_recursion_scheduled[get_configuration1-/test_recursion_320-subdir-318] SKIPPED     [ 54%]
test_recursion_level.py::test_recursion_scheduled[get_configuration1-/test recursion 320-sub dir -318] SKIPPED   [ 55%]
test_recursion_level.py::test_recursion_whodata[get_configuration1-/test_no_recursion-subdir-0] SKIPPED          [ 56%]
test_recursion_level.py::test_recursion_whodata[get_configuration1-/test no recursion-sub dir -0] SKIPPED        [ 58%]
test_recursion_level.py::test_recursion_whodata[get_configuration1-/test_recursion_1-subdir-1] SKIPPED           [ 59%]
test_recursion_level.py::test_recursion_whodata[get_configuration1-/test recursion 1-sub dir -1] SKIPPED         [ 61%]
test_recursion_level.py::test_recursion_whodata[get_configuration1-/test_recursion_5-subdir-5] SKIPPED           [ 62%]
test_recursion_level.py::test_recursion_whodata[get_configuration1-/test recursion 5-sub dir -5] SKIPPED         [ 63%]
test_recursion_level.py::test_recursion_whodata[get_configuration1-/test_recursion_320-subdir-318] SKIPPED       [ 65%]
test_recursion_level.py::test_recursion_whodata[get_configuration1-/test recursion 320-sub dir -318] SKIPPED     [ 66%]
test_recursion_level.py::test_recursion_realtime[get_configuration2-/test_no_recursion-subdir-0] SKIPPED         [ 68%]
test_recursion_level.py::test_recursion_realtime[get_configuration2-/test no recursion-sub dir -0] SKIPPED       [ 69%]
test_recursion_level.py::test_recursion_realtime[get_configuration2-/test_recursion_1-subdir-1] SKIPPED          [ 70%]
test_recursion_level.py::test_recursion_realtime[get_configuration2-/test recursion 1-sub dir -1] SKIPPED        [ 72%]
test_recursion_level.py::test_recursion_realtime[get_configuration2-/test_recursion_5-subdir-5] SKIPPED          [ 73%]
test_recursion_level.py::test_recursion_realtime[get_configuration2-/test recursion 5-sub dir -5] SKIPPED        [ 75%]
test_recursion_level.py::test_recursion_realtime[get_configuration2-/test_recursion_320-subdir-318] SKIPPED      [ 76%]
test_recursion_level.py::test_recursion_realtime[get_configuration2-/test recursion 320-sub dir -318] SKIPPED    [ 77%]
test_recursion_level.py::test_recursion_scheduled[get_configuration2-/test_no_recursion-subdir-0] SKIPPED        [ 79%]
test_recursion_level.py::test_recursion_scheduled[get_configuration2-/test no recursion-sub dir -0] SKIPPED      [ 80%]
test_recursion_level.py::test_recursion_scheduled[get_configuration2-/test_recursion_1-subdir-1] SKIPPED         [ 81%]
test_recursion_level.py::test_recursion_scheduled[get_configuration2-/test recursion 1-sub dir -1] SKIPPED       [ 83%]
test_recursion_level.py::test_recursion_scheduled[get_configuration2-/test_recursion_5-subdir-5] SKIPPED         [ 84%]
test_recursion_level.py::test_recursion_scheduled[get_configuration2-/test recursion 5-sub dir -5] SKIPPED       [ 86%]
test_recursion_level.py::test_recursion_scheduled[get_configuration2-/test_recursion_320-subdir-318] SKIPPED     [ 87%]
test_recursion_level.py::test_recursion_scheduled[get_configuration2-/test recursion 320-sub dir -318] SKIPPED   [ 88%]
test_recursion_level.py::test_recursion_whodata[get_configuration2-/test_no_recursion-subdir-0] PASSED           [ 90%]
test_recursion_level.py::test_recursion_whodata[get_configuration2-/test no recursion-sub dir -0] PASSED         [ 91%]
test_recursion_level.py::test_recursion_whodata[get_configuration2-/test_recursion_1-subdir-1] PASSED            [ 93%]
test_recursion_level.py::test_recursion_whodata[get_configuration2-/test recursion 1-sub dir -1] PASSED          [ 94%]
test_recursion_level.py::test_recursion_whodata[get_configuration2-/test_recursion_5-subdir-5] PASSED            [ 95%]
test_recursion_level.py::test_recursion_whodata[get_configuration2-/test recursion 5-sub dir -5] PASSED          [ 97%]
test_recursion_level.py::test_recursion_whodata[get_configuration2-/test_recursion_320-subdir-318] FAILED        [ 98%]
test_recursion_level.py::test_recursion_whodata[get_configuration2-/test recursion 320-sub dir -318] FAILED      [100%]

================================ 2 failed, 14 passed, 56 skipped in 300.06s (0:05:00) =================================
```

Two Whodata tests are currently failing due to an incosistency when using very long paths. This issue has been reported and its outside the scope of this tests

Best regards,
Carlos.

### CODE REVIEW

File: packages/wazuh_testing/wazuh_testing/fim.py [hunk 1]
```diff
@@ -268,7 +269,8 @@ def change_internal_options(opt_path, pattern, value):
 
     with open(opt_path, "w") as sources:
         for line in lines:
-            sources.write(re.sub(f'{pattern}=[0-9]*', f'{pattern}={value}', line))
+            sources.write(
+                re.sub(f'{pattern}=[0-9]*', f'{pattern}={value}', line))
```
Comment (crd1985) [inline packages/wazuh_testing/wazuh_testing/fim.py]: I would not apply this change. If you want to improve readability create a variable above with the returned value of re.sub.

File: packages/wazuh_testing/wazuh_testing/fim.py [hunk 2]
```diff
@@ -337,87 +339,88 @@ def callback_configuration_error(line):
...
-                               accum_results=n_regular).result()
-
-    # Are the n_regular events of type 'deleted'?
-    type_assert('deleted')
-    check_files_in_event()
+    check_time_travel()
+    events = fetch_events()
+    if events is not None:
+        check_events_type('deleted')
+        check_files_in_event()
```
Comment (crd1985) [inline packages/wazuh_testing/wazuh_testing/fim.py]: No newline at end of file.

File: test_wazuh/test_fim/test_recursion_level/data/wazuh_conf.yaml [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      attributes:
+      - check_all: 'yes'
+      - whodata: 'yes'
+      - recursion_level: "320"
+  - directories:
+      value: "/test recursion 320"
+      attributes:
+      - check_all: 'yes'
+      - whodata: 'yes'
+      - recursion_level: "320"
```
Comment (crd1985) [inline test_wazuh/test_fim/test_recursion_level/data/wazuh_conf.yaml]: No new line at end of file

Comment (crd1985) [inline test_wazuh/test_fim/test_recursion_level/data/wazuh_conf.yaml]: Maybe the configuration could be defined once and be parametrized using the `params` and `metadata` parameters of `load_wazuh_configurations` function.

File: test_wazuh/test_fim/test_basic_usage/test_basic_usage_changes.py [hunk 4]
```diff
@@ -61,4 +61,4 @@ def test_regular_file_changes(folder, checkers, is_scheduled, tags_to_apply,
     n_regular = 3
     min_timeout = 3
 
-    regular_file_cud(folder, is_scheduled, n_regular, min_timeout, wazuh_log_monitor, checkers)
+    regular_file_cud(folder, wazuh_log_monitor, is_scheduled, n_regular, min_timeout, checkers)
```
Comment (crd1985) [inline test_wazuh/test_fim/test_basic_usage/test_basic_usage_changes.py]:
```suggestion
    regular_file_cud(folder, wazuh_log_monitor, time_travel=is_scheduled, n_regular=n_regular, min_timeout=min_timeout, options=checkers)
```

File: test_wazuh/test_fim/test_benchmark/test_benchmark.py [hunk 5]
```diff
@@ -50,5 +50,4 @@ def test_benchmark_regular_files(n_regular, folder, is_scheduled,
     check_apply_test(tags_to_apply, get_configuration['tags'])
     min_timeout = 30
 
-    regular_file_cud(folder, is_scheduled, n_regular, min_timeout,
-                     wazuh_log_monitor)
+    regular_file_cud(folder, wazuh_log_monitor, is_scheduled, n_regular, min_timeout)
```
Comment (crd1985) [inline test_wazuh/test_fim/test_benchmark/test_benchmark.py]: Same as before.

File: packages/wazuh_testing/wazuh_testing/fim.py (hunk 1 above)
Comment (CarlosRS9) [reply packages/wazuh_testing/wazuh_testing/fim.py]: Fixed. That line was modified by mistake. The newline there was not intended.

File: packages/wazuh_testing/wazuh_testing/fim.py (hunk 2 above)
Comment (CarlosRS9) [reply packages/wazuh_testing/wazuh_testing/fim.py]: New line added at end of file.

File: test_wazuh/test_fim/test_recursion_level/data/wazuh_conf.yaml (hunk 3 above)
Comment (CarlosRS9) [reply test_wazuh/test_fim/test_recursion_level/data/wazuh_conf.yaml]: New line added at end of file.

Comment (CarlosRS9) [reply test_wazuh/test_fim/test_recursion_level/data/wazuh_conf.yaml]: The config file has been refactorized to use `params` and `metadata` as suggested. The necessary modifications have been made to the test function to correctly use this new configuration file.

File: test_wazuh/test_fim/test_basic_usage/test_basic_usage_changes.py (hunk 4 above)
Comment (CarlosRS9) [reply test_wazuh/test_fim/test_basic_usage/test_basic_usage_changes.py]: Done

File: test_wazuh/test_fim/test_benchmark/test_benchmark.py (hunk 5 above)
Comment (CarlosRS9) [reply test_wazuh/test_fim/test_benchmark/test_benchmark.py]: Done

**Comment by CarlosRS9** [comment]
All requested changes has been applied and the test functions for realtime, whodata and scheduled has been unified in a single test function.

The test affected by those modifications has been executed again to ensure that they did not have any impact. Everything is working as expected.

**Pull request merged**